test/rsa_sp800_56b_test.c

```diff
@@ -300,9 +300,17 @@ static int test_check_private_exponent(void)
           && TEST_true(BN_set_word(d, 157))
           && TEST_true(BN_set_word(n, 15*17))
           && TEST_true(RSA_set0_factors(key, p, q))
-          && TEST_true(RSA_set0_key(key, n, e, d))
-          /* fails since d >= lcm(p-1, q-1) */
-          && TEST_false(rsa_check_private_exponent(key, 8, ctx))
+          && TEST_true(RSA_set0_key(key, n, e, d));
+    if (!ret) {
+        BN_free(p);
+        BN_free(q);
+        BN_free(e);
+        BN_free(d);
+        BN_free(n);
+        goto end;
+    }
+    /* fails since d >= lcm(p-1, q-1) */
+    ret = TEST_false(rsa_check_private_exponent(key, 8, ctx))
```

Comment:
I don't understand why you don't keep doing &&. If any of the previous tests failed, it would just skip all the other tests.

Reply:
The tests all pass. 
I am not sure breaking it up into seperate checks is really going to help much.
The code I have seen in other tests does the following pattern everywhere:
 if (fail) goto err;
 
If any error appears, then the normal process would be to build the test locally and fix any problems. Failing multiple times might help if there is a platform specific problem where it is harder to reproduce, but for generic code I dont think you gain much advantage by making the test more complex.

Reply:
I think you misunderstand me. If you do A() && B(), and A() fails, then B() does not get run. So I see no point in adding an if (!ret) in the middle of that, and then just continue with ret = D() && E() && ...

Reply:
Could you show me how you would write the code in test_check_private_exponent()?
Keeping in mind that the RSA_set0_key(n,e,d) takes ownership of n,e,d.
If the RSA_set0_key fails then n,e & d need to use BN_free. If the set succeeds then RSA_free() will free up the n,e,d resources.

Reply:
Oh, I forgot all about the ownership transfer

Reply:
things get ugly when someone else grabs ownership :)